### DISCUSSION TITLE
Trigger async data loads reactively instead of in an action

### DISCUSSION THREAD

**Discussion by jheeffer** [mobxjs/mobx#3007] [Q&A]
All the examples of async data loading I found trigger the load inside an action, the [docs ](https://mobx.js.org/actions.html#asynchronous-actions) being the most obvious one.

However, I'd like a pattern to trigger a data load reactively. In other words, the trigger is derived from base or computed state. For example, I change which database field I want to fetch, a `computed` state creates an SQL query, which thén triggers an async data load through a `reaction` or a `computed`. Or the same without the `computed` intermediate step.

Sure, I could create an `action` that creates the query and triggers the load, and call it from every `action` changing base state (like the database field). But the whole idea of mobx is that I wouldn't have to. Whether you consider async data (loads) side-effects or derived async state, it should be possible to trigger it reactively. Moreover, I wouldn't be able to use mobx' strength of resolving complex state-derivation trees to construct the query.

The two ways to do this reactively in mobx would be with either a `computed` or a `reaction` (incl `autorun`, `when`, etc).

None of my ideas work exactly like I want to. Requirements I have trouble with are: 
 1. Always fetch áfter the outermost action is finished, even if action reads async state, so that the async request isn't triggered with inconsistent state.
 2. Fetch & set state to "loading" befóre any external reactions runs (i.e. the application won't react before async request is done and my model state is thus consistent)
 3. Be fully reactive: no need to call extra actions after changing state and no need to keep track of exactly what observable props cóuld trigger an async request.

I have a extensive documentation of the patterns I came up with and their drawbacks here: 
https://github.com/vizabi/vizabi-reactive/blob/develop/docs/data%20loading.md

I'll try to summarize here:

- Fetch in a computed, returning a `fromPromise()`. State & value computed props reflect the fromPromise. Doesn't fulfill 1 as computed can get recomputed during action. [Example JS Fiddle](https://jsfiddle.net/jasperh/2h6f5dv3/12/).
- Fetch in an autorun, setting a `fromPromise()` to an observable prop. State & value computed props reflect the fromPromise. Doesn't fulfill 2 as external reaction can run before fetch-autorun. [Example JS Fiddle](https://jsfiddle.net/jasperh/8c7mbLsj/75/).
- Fetch in an autorun, but keep the action's updates away from application until async resolves. Doesn't fulfill 3 as you need to keep track of which updates to "keep away from application". And complicates code a lot. [Example JS Fiddle](https://jsfiddle.net/jasperh/oy8bae1z/2/).

Are there any known patterns for this which fulfill all 3 requirements?

If we can come up with a solution, I'd love to make a PR to improve the docs with examples of a useful pattern! Or document why this is an anti-pattern, if it is, and what the alternative is.

**Comment by mweststrate** [comment]
`reaction` allows you to precisely define what you want to react to for fetching, instead of keeping it generic. In general remember that you don't have to apply your state updates immedaitely when doing a fetch, you can still 'batch' them, for example by storing the intended updates in local variables or a closure, and apply them at the end, if you don't want the app to react.

Beyond that I find the examples a bit hard to grasp as they are too abstract. Could you translate the question to a simple but realistic case that is a bit easier to reason about?

**Reply by jheeffer** [reply]
>  In general remember that you don't have to apply your state updates immedaitely when doing a fetch, you can still 'batch' them, for example by storing the intended updates in local variables or a closure, and apply them at the end, if you don't want the app to react.

Yep, that's my 3rd example (just updated, I had a wrong link before, which you might have seen). The code in my example seems convoluted to me.
Also, not included in this example yet, it would mean I'd have to clone my store to trigger the async load and then apply them to the "real" store when I want the app to react.

I'll try to come up with a more practical use case.

**Reply by jheeffer** [reply]
> reaction allows you to precisely define what you want to react to for fetching, 

Yep! That's what I like about the autorun or computed patterns, they keep track of precisely the observables needed for the fetch, however they have their own drawbacks : ).

**Reply by jheeffer** [reply]
https://jsfiddle.net/jasperh/2h6f5dv3/30/

Here's a more practical example of the second pattern: using an autorun to trigger the async load.

**Reply by jheeffer** [reply]
https://jsfiddle.net/jasperh/2h6f5dv3/53/

Here's the same example but without immediately applying my state updates (application side) when doing a fetch, and instead postponing update untill after the fetch response.
I'm not sure how to do this without copying the full store's base & computed fields to a second one. I need two states right? One for the application, one internally to send the right query.

**Reply by jheeffer** [reply]
I was able to clean it up significantly using `mobxUtils.createViewModel`.

https://jsfiddle.net/jasperh/2h6f5dv3/62/

However, actions writing to the model now should know if a state change triggers async load (then write to viewModel) or not (then write directly to model). That's... sub-optimal, requiring implementation knowledge to use the API correctly.

**Reply by jheeffer** [reply]
Here's an example of non-async-triggering state which needs to be set to model instead of viewModel (constant value that has priority over query).

https://jsfiddle.net/jasperh/2h6f5dv3/69/

**Reply by jheeffer** [reply]
I updated the question to hopefully make it simpler. In a nutshell: "How to trigger async data requests reactively?"
